### PR TITLE
[rllib]Add entropy coeff decay

### PR DESCRIPTION
**Pull request by stefanpantic**
## What do these changes do?
* Adds an entropy coefficient decay schedule API
* Add entropy coefficient decay to IMPALA, APPO and PPO

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/rllib/policy/tf_policy.py [hunk 1]
```diff
@@ -544,3 +544,24 @@ def on_global_var_update(self, global_vars):
...
+        self._entropy_schedule = entropy_coeff_schedule
+
+    @override(Policy)
+    def on_global_var_update(self, global_vars):
+        super(EntropyCoeffSchedule, self).on_global_var_update(global_vars)
+        if self._entropy_schedule is not None:
+            self.entropy_coeff.load(
+                self.entropy_coeff.eval(session=self._sess) *
+                (1 - global_vars['timestep'] / self.
+                 config['entropy_coeff_schedule']),
```
Comment (ericl) [inline python/ray/rllib/policy/tf_policy.py]: nit: use double quotes

**Review by ericl — APPROVED** [review]
Looks good. I think you might have to install yapf==0.23.0 though to get the same formatting as the repo does.

**Comment by ericl** [comment]
```
$ flake8 --inline-quotes '"' --no-avoid-escape --exclude=python/ray/core/generated/,doc/source/conf.py,python/ray/cloudpickle/ --ignore=C408,E121,E123,E126,E226,E24,E704,W503,W504,W605
./python/ray/autoscaler/autoscaler.py:673:80: E501 line too long (92 > 79 characters)
./python/ray/rllib/agents/ppo/ppo_policy.py:11:80: E501 line too long (81 > 79 characters)
./python/ray/rllib/agents/ppo/appo_policy.py:19:80: E501 line too long (81 > 79 characters)
The command "flake8 --inline-quotes '"' --no-avoid-escape --exclude=python/ray/core/generated/,doc/source/conf.py,python/ray/cloudpickle/ --ignore=C408,E121,E123,E126,E226,E24,E704,W503,W504,W605" exited with 1.
43.22s$ ./ci/travis/format.sh --all
From https://github.com/ray-project/ray
 * branch            master     -> FETCH_HEAD
 * [new branch]      master     -> upstream/master
--- python/ray/autoscaler/autoscaler.py	(original)
+++ python/ray/autoscaler/autoscaler.py	(reformatted)
@@ -670,7 +670,9 @@
                 nodes = self.workers()
                 if nodes:
                     self.provider.terminate_nodes(nodes)
-                logger.error("StandardAutoscaler: terminated {} node(s)".format(len(nodes)))
+                logger.error(
+                    "StandardAutoscaler: terminated {} node(s)".format(
+                        len(nodes)))
             except Exception:
                 traceback.print_exc()
 
The command "./ci/travis/format.sh --all" exited with 1.
```

(from travis)

**Comment by ericl** [comment]
@stefanpantic I merged in the new vtrace policy refactoring, let me know if it still looks good

**Comment by stefanpantic** [comment]
> @stefanpantic I merged in the new vtrace policy refactoring, let me know if it still looks good
Looks good. Everyghing works as expected.

image

**Comment by yangysc** [comment]
> > @stefanpantic I merged in the new vtrace policy refactoring, let me know if it still looks good
> > Looks good. Everyghing works as expected.
> 
> image

May I ask that how do you set the _entropy_coeff_schedule_?  It seems to be the maximum of the total time step, which is unknown usually.  Why is the api of entropy schedule not the same with the lr schedule?

File: python/ray/rllib/policy/tf_policy.py [hunk 2]
```diff
@@ -565,3 +565,24 @@ def on_global_var_update(self, global_vars):
...
+        self._entropy_schedule = entropy_coeff_schedule
+
+    @override(Policy)
+    def on_global_var_update(self, global_vars):
+        super(EntropyCoeffSchedule, self).on_global_var_update(global_vars)
+        if self._entropy_schedule is not None:
+            self.entropy_coeff.load(
+                self.entropy_coeff.eval(session=self._sess) *
+                (1 - global_vars["timestep"] /
+                 self.config["entropy_coeff_schedule"]),
```
Comment (ericl) [inline python/ray/rllib/policy/tf_policy.py]: @stefanpantic doesn't this break if the timestep value becomes greater than the schedule? cc @dmadeka

Comment (stefanpantic) [reply python/ray/rllib/policy/tf_policy.py]: It shouldn’t. It will just decay to 0 at that point. I’ll test later and get back to you.

**Pull request merged**